File: import_off.js
#!/usr/bin/env node
'use strict';

const fs = require('fs');
const path = require('path');
const https = require('https');
const http = require('http');
const zlib = require('zlib');
const readline = require('readline');
const Database = require('better-sqlite3');

const DATA_DIR = process.env.DATA_DIR || '/data';
const OFF_DB = process.env.OFF_DB || path.join(DATA_DIR, 'off.db');
const LOG_FILE = path.join(DATA_DIR, 'off_update.log');
const OFF_URL = process.env.OFF_SOURCE_URL || 'https://static.openfoodfacts.org/data/en.openfoodfacts.org.products.csv.gz';
const OFF_INPUT = process.env.OFF_INPUT || '';
const LIMIT = Number(process.env.OFF_IMPORT_LIMIT || '0') || 0;
const TMP_TABLE = 'aliments_import_tmp';

function ensureDir(dir) { if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true }); }
ensureDir(DATA_DIR);

function log(msg) {
  const line = `[${new Date().toISOString()}] ${msg}`;
  console.log(line);
  fs.appendFileSync(LOG_FILE, line + '\n');
}

function val(row, name) { return row[name] == null ? '' : String(row[name]).trim(); }
function num(row, ...names) {
  for (const name of names) {
    const raw = val(row, name).replace(',', '.');
    if (raw === '') continue;
    const n = Number(raw);
    if (Number.isFinite(n)) return n;
  }
  return null;
}
function kcal(row) {
  const k = num(row, 'energy-kcal_100g', 'energy-kcal_value_100g', 'kcal100');
  if (k != null && k > 0) return k;
  const kj = num(row, 'energy-kj_100g', 'energy_100g');
  if (kj != null && kj > 0) return Math.round((kj / 4.184) * 10) / 10;
  return null;
}
function pickName(row) {
  return val(row, 'product_name_fr') || val(row, 'product_name') || val(row, 'generic_name_fr') || val(row, 'generic_name');
}
function makeRow(headers, parts) {
  const row = {};
  for (let i = 0; i < headers.length; i++) row[headers[i]] = parts[i] || '';
  return row;
}
function normalizeCode(code) {
  const c = String(code || '').trim().replace(/[^0-9A-Za-z_-]/g, '');
  return c || null;
}

function download(url, redirectCount = 0) {
  return new Promise((resolve, reject) => {
    const lib = url.startsWith('https:') ? https : http;
    const req = lib.get(url, { headers: { 'User-Agent': 'FoodNote self-hosted importer/10.49' } }, (res) => {
      if (res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
        if (redirectCount > 5) return reject(new Error('Trop de redirections OpenFoodFacts'));
        const nextUrl = new URL(res.headers.location, url).toString();
        log(`Redirection OpenFoodFacts: ${nextUrl}`);
        res.resume();
        return resolve(download(nextUrl, redirectCount + 1));
      }
      if (res.statusCode !== 200) {
        res.resume();
        return reject(new Error(`HTTP ${res.statusCode} lors du téléchargement OpenFoodFacts`));
      }
      resolve(url.endsWith('.gz') ? res.pipe(zlib.createGunzip()) : res);
    });
    req.on('error', reject);
  });
}

async function openInputStream() {
  if (OFF_INPUT) {
    log(`Lecture fichier local: ${OFF_INPUT}`);
    const stream = fs.createReadStream(OFF_INPUT);
    return OFF_INPUT.endsWith('.gz') ? stream.pipe(zlib.createGunzip()) : stream;
  }
  log(`Téléchargement OpenFoodFacts: ${OFF_URL}`);
  return download(OFF_URL);
}

function createImportTable(db) {
  db.exec(`
    DROP TABLE IF EXISTS ${TMP_TABLE};
    CREATE TABLE ${TMP_TABLE} (
      id INTEGER PRIMARY KEY,
      code TEXT,
      nom TEXT NOT NULL,
      marque TEXT,
      kcal100 REAL,
      prot100 REAL,
      gluc100 REAL,
      lip100 REAL,
      fibres100 REAL
    );
    CREATE INDEX IF NOT EXISTS idx_${TMP_TABLE}_code ON ${TMP_TABLE}(code);
    CREATE INDEX IF NOT EXISTS idx_${TMP_TABLE}_nom ON ${TMP_TABLE}(nom);
  `);
}

function replaceOfficialTable(db) {
  db.exec(`
    DROP TABLE IF EXISTS aliments_old_before_barcode;
    ALTER TABLE aliments RENAME TO aliments_old_before_barcode;
    ALTER TABLE ${TMP_TABLE} RENAME TO aliments;
    DROP TABLE IF EXISTS aliments_old_before_barcode;
    CREATE INDEX IF NOT EXISTS idx_off_aliments_code ON aliments(code);
    CREATE INDEX IF NOT EXISTS idx_off_aliments_nom ON aliments(nom);
    ANALYZE;
  `);
}

async function main() {
  fs.writeFileSync(LOG_FILE, '');
  log('Import OpenFoodFacts v10.49 — code-barres + import robuste gros champs');
  log(`Base cible: ${OFF_DB}`);
  log(`Source: ${OFF_INPUT || OFF_URL}`);

  const db = new Database(OFF_DB);
  db.pragma('journal_mode = WAL');
  db.pragma('synchronous = NORMAL');
  db.exec(`
    CREATE TABLE IF NOT EXISTS aliments (
      id INTEGER PRIMARY KEY,
      code TEXT,
      nom TEXT NOT NULL,
      marque TEXT,
      kcal100 REAL,
      prot100 REAL,
      gluc100 REAL,
      lip100 REAL,
      fibres100 REAL
    );
  `);
  createImportTable(db);

  const insert = db.prepare(`
    INSERT INTO ${TMP_TABLE} (code, nom, marque, kcal100, prot100, gluc100, lip100, fibres100)
    VALUES (?, ?, ?, ?, ?, ?, ?, ?)
  `);
  const tx = db.transaction((items) => { for (const it of items) insert.run(...it); });

  let inserted = 0, skipped = 0, withCode = 0, lineNo = 0, rowsRead = 0;
  let batch = [];
  let headers = null;
  let hasCodeHeader = false;
  let codeHeaderName = null;

  const input = await openInputStream();
  const rl = readline.createInterface({ input, crlfDelay: Infinity });

  for await (const line of rl) {
    lineNo++;
    if (!headers) {
      headers = line.split('\t').map(h => h.trim());
      codeHeaderName = ['code', 'barcode', 'ean', 'product_code'].find(h => headers.includes(h)) || null;
      hasCodeHeader = !!codeHeaderName;
      log(`Colonnes détectées: ${headers.length}`);
      log(`Premières colonnes: ${headers.slice(0, 25).join(', ')}`);
      log(`Colonne code-barres source: ${codeHeaderName || 'ABSENTE'}`);
      if (!hasCodeHeader) {
        db.exec(`DROP TABLE IF EXISTS ${TMP_TABLE}`);
        db.close();
        throw new Error('La source OpenFoodFacts ne contient aucune colonne code/barcode/ean/product_code. Import annulé pour ne pas remplacer la base par une version non scannable.');
      }
      continue;
    }
    rowsRead++;
    if (!line.trim()) continue;
    const parts = line.split('\t');
    const row = makeRow(headers, parts);
    const code = normalizeCode(val(row, codeHeaderName));
    const nom = pickName(row);
    const k = kcal(row);
    if (!nom || !k || k <= 0 || k >= 1000) { skipped++; continue; }
    const item = [
      code,
      nom.slice(0, 240),
      (val(row, 'brands') || val(row, 'marque')).split(',')[0].slice(0, 160) || null,
      k,
      num(row, 'proteins_100g', 'prot100'),
      num(row, 'carbohydrates_100g', 'gluc100'),
      num(row, 'fat_100g', 'lip100'),
      num(row, 'fiber_100g', 'fibres100')
    ];
    if (code) withCode++;
    batch.push(item);
    inserted++;
    if (batch.length >= 2000) { tx(batch); batch = []; }
    if (inserted % 50000 === 0) log(`Importés: ${inserted} produits (${withCode} avec code-barres, ${skipped} ignorés)`);
    if (LIMIT && inserted >= LIMIT) break;
  }
  if (batch.length) tx(batch);

  const tmpTotal = db.prepare(`SELECT COUNT(*) AS n FROM ${TMP_TABLE}`).get().n;
  const tmpWithCode = db.prepare(`SELECT COUNT(*) AS n FROM ${TMP_TABLE} WHERE code IS NOT NULL AND TRIM(code) != ''`).get().n;
  log(`Table temporaire terminée: ${tmpTotal} produits, ${tmpWithCode} avec code-barres, ${skipped} ignorés, ${rowsRead} lignes lues.`);

  if (tmpTotal === 0) {
    db.exec(`DROP TABLE IF EXISTS ${TMP_TABLE}`);
    db.close();
    throw new Error('Import annulé: aucun produit exploitable trouvé. Ancienne base conservée.');
  }
  if (tmpWithCode === 0) {
    db.exec(`DROP TABLE IF EXISTS ${TMP_TABLE}`);
    db.close();
    throw new Error('Import annulé: 0 produit avec code-barres. Ancienne base conservée. Vérifie la source OpenFoodFacts utilisée.');
  }

  replaceOfficialTable(db);
  const cols = db.prepare('PRAGMA table_info(aliments)').all().map(c => c.name).join(', ');
  const total = db.prepare('SELECT COUNT(*) AS n FROM aliments').get().n;
  const barcode = db.prepare("SELECT COUNT(*) AS n FROM aliments WHERE code IS NOT NULL AND TRIM(code) != ''").get().n;
  db.close();
  log(`Schéma final aliments: ${cols}`);
  log(`Terminé. Produits: ${total}. Avec code-barres: ${barcode}. Sans code-barres: ${Math.max(0, total - barcode)}. Ignorés: ${skipped}.`);
}

main().catch(err => { log(`ERREUR: ${err.stack || err.message}`); process.exit(1); });

File: import_off.py
#!/usr/bin/env python3
"""
FoodNote v10.50 — import OpenFoodFacts compatible Python/Node.

But : éviter définitivement :
    _csv.Error: field larger than field limit (131072)

Priorité : utiliser import_off.js, qui lit le dump OpenFoodFacts en streaming et
conserve le champ `code` pour le scan code-barres local.

Si Node/import_off.js est indisponible, ce fichier Python peut quand même importer
le TSV/CSV avec une limite CSV fortement augmentée.
"""
import csv
import gzip
import os
import sqlite3
import subprocess
import sys
import urllib.request
from pathlib import Path

# Corrige le vrai plantage Python : limite CSV par défaut trop basse.
try:
    limit = sys.maxsize
    while True:
        try:
            csv.field_size_limit(limit)
            break
        except OverflowError:
            limit = int(limit / 10)
except Exception:
    csv.field_size_limit(1024 * 1024 * 1024)

DATA_DIR = Path(os.environ.get("DATA_DIR", "/data"))
OFF_DB = Path(os.environ.get("OFF_DB", str(DATA_DIR / "off.db")))
OFF_URL = os.environ.get("OFF_SOURCE_URL", "https://static.openfoodfacts.org/data/en.openfoodfacts.org.products.csv.gz")
OFF_INPUT = os.environ.get("OFF_INPUT", "")
LIMIT = int(os.environ.get("OFF_IMPORT_LIMIT", "0") or "0")
LOG_FILE = DATA_DIR / "off_update.log"


def log(msg):
    DATA_DIR.mkdir(parents=True, exist_ok=True)
    line = str(msg)
    print(line, flush=True)
    with LOG_FILE.open("a", encoding="utf-8") as f:
        f.write(line + "\n")


def run_node_import_if_available():
    script = Path(__file__).with_name("import_off.js")
    if not script.exists():
        script = Path("/app/import_off.js")
    if script.exists():
        log("FoodNote v10.50: import OpenFoodFacts via import_off.js")
        return subprocess.call(["node", str(script)], env=os.environ.copy())
    return None


def clean_num(v):
    if v is None:
        return None
    s = str(v).strip().replace(",", ".")
    if not s:
        return None
    try:
        n = float(s)
        if n != n:
            return None
        return n
    except Exception:
        return None


def first(row, *names):
    for name in names:
        v = row.get(name, "")
        if v is not None and str(v).strip():
            return str(v).strip()
    return ""


def kcal(row):
    k = clean_num(first(row, "energy-kcal_100g", "energy-kcal_value_100g", "kcal100"))
    if k and k > 0:
        return k
    kj = clean_num(first(row, "energy-kj_100g", "energy_100g"))
    if kj and kj > 0:
        return round(kj / 4.184, 1)
    return None


def open_source():
    if OFF_INPUT:
        p = Path(OFF_INPUT)
        log(f"Lecture source locale OpenFoodFacts: {p}")
        raw = p.open("rb")
        return gzip.open(raw, "rt", encoding="utf-8", errors="replace", newline="") if p.suffix == ".gz" else open(p, "r", encoding="utf-8", errors="replace", newline="")
    log(f"Téléchargement OpenFoodFacts: {OFF_URL}")
    resp = urllib.request.urlopen(OFF_URL, timeout=120)
    return gzip.open(resp, "rt", encoding="utf-8", errors="replace", newline="") if OFF_URL.endswith(".gz") else resp


def build_db():
    DATA_DIR.mkdir(parents=True, exist_ok=True)
    LOG_FILE.write_text("", encoding="utf-8")
    log("FoodNote v10.50: import Python robuste OpenFoodFacts")
    log(f"Base cible: {OFF_DB}")

    con = sqlite3.connect(str(OFF_DB))
    cur = con.cursor()
    cur.execute("PRAGMA journal_mode=WAL")
    cur.execute("PRAGMA synchronous=NORMAL")
    cur.executescript("""
    DROP TABLE IF EXISTS aliments_import_tmp;
    CREATE TABLE aliments_import_tmp (
      id INTEGER PRIMARY KEY,
      code TEXT,
      nom TEXT NOT NULL,
      marque TEXT,
      kcal100 REAL,
      prot100 REAL,
      gluc100 REAL,
      lip100 REAL,
      fibres100 REAL
    );
    CREATE INDEX IF NOT EXISTS idx_aliments_import_tmp_code ON aliments_import_tmp(code);
    CREATE INDEX IF NOT EXISTS idx_aliments_import_tmp_nom ON aliments_import_tmp(nom);
    """)
    con.commit()

    inserted = 0
    with_code = 0
    skipped = 0
    rows_read = 0
    batch = []

    with open_source() as f:
        reader = csv.DictReader(f, delimiter="\t")
        headers = reader.fieldnames or []
        code_col = next((x for x in ["code", "barcode", "ean", "product_code"] if x in headers), None)
        log(f"Colonnes détectées: {len(headers)}")
        log("Premières colonnes: " + ", ".join(headers[:25]))
        log("Colonne code-barres source: " + (code_col or "ABSENTE"))
        if not code_col:
            cur.execute("DROP TABLE IF EXISTS aliments_import_tmp")
            con.commit(); con.close()
            raise SystemExit("ERREUR: aucune colonne code/barcode/ean/product_code. Import annulé.")

        for row in reader:
            rows_read += 1
            name = first(row, "product_name_fr", "product_name", "generic_name_fr", "generic_name")
            k = kcal(row)
            if not name or not k or k <= 0 or k >= 1000:
                skipped += 1
                continue
            code = first(row, code_col)
            code = "".join(ch for ch in code if ch.isalnum() or ch in "_-") or None
            if code:
                with_code += 1
            marque = first(row, "brands", "marque").split(",")[0][:160] or None
            batch.append((code, name[:240], marque, k,
                          clean_num(first(row, "proteins_100g", "prot100")),
                          clean_num(first(row, "carbohydrates_100g", "gluc100")),
                          clean_num(first(row, "fat_100g", "lip100")),
                          clean_num(first(row, "fiber_100g", "fibres100"))))
            inserted += 1
            if len(batch) >= 2000:
                cur.executemany("""
                  INSERT INTO aliments_import_tmp
                  (code, nom, marque, kcal100, prot100, gluc100, lip100, fibres100)
                  VALUES (?, ?, ?, ?, ?, ?, ?, ?)
                """, batch)
                con.commit(); batch.clear()
            if inserted and inserted % 50000 == 0:
                log(f"Importés: {inserted} produits ({with_code} avec code-barres, {skipped} ignorés)")
            if LIMIT and inserted >= LIMIT:
                break

    if batch:
        cur.executemany("""
          INSERT INTO aliments_import_tmp
          (code, nom, marque, kcal100, prot100, gluc100, lip100, fibres100)
          VALUES (?, ?, ?, ?, ?, ?, ?, ?)
        """, batch)
        con.commit()

    total = cur.execute("SELECT COUNT(*) FROM aliments_import_tmp").fetchone()[0]
    bc = cur.execute("SELECT COUNT(*) FROM aliments_import_tmp WHERE code IS NOT NULL AND TRIM(code) != ''").fetchone()[0]
    log(f"Table temporaire: {total} produits, {bc} avec code-barres, {skipped} ignorés, {rows_read} lignes lues")
    if total <= 0 or bc <= 0:
        cur.execute("DROP TABLE IF EXISTS aliments_import_tmp")
        con.commit(); con.close()
        raise SystemExit("ERREUR: import annulé, aucun produit exploitable ou aucun code-barres. Ancienne base conservée.")

    cur.executescript("""
    DROP TABLE IF EXISTS aliments_old_before_barcode;
    ALTER TABLE aliments RENAME TO aliments_old_before_barcode;
    ALTER TABLE aliments_import_tmp RENAME TO aliments;
    DROP TABLE IF EXISTS aliments_old_before_barcode;
    CREATE INDEX IF NOT EXISTS idx_off_aliments_code ON aliments(code);
    CREATE INDEX IF NOT EXISTS idx_off_aliments_nom ON aliments(nom);
    ANALYZE;
    """)
    con.commit(); con.close()
    log(f"Terminé. Produits: {total}. Avec code-barres: {bc}. Sans code-barres: {max(0,total-bc)}.")


if __name__ == "__main__":
    # Par défaut, on privilégie le nouvel import Node. Si OFF_FORCE_PYTHON=1,
    # on utilise l'import Python robuste ci-dessus.
    if os.environ.get("OFF_FORCE_PYTHON") != "1":
        code = run_node_import_if_available()
        if code is not None:
            raise SystemExit(code)
    build_db()
